## **Maestría en Inteligencia Artificial Aplicada**
### **Curso: Inteligencia Artificial y Aprendizaje Automático**
#### Tecnológico de Monterrey
#### Prof Luis Eduardo Falcón Morales

### **Adtividad de la Semana 8**
#### **Reducción de Dimensionalidad - Descomposición en Valores Singulares (SVD) y Sistemas de Recomendación**


**Esta Actividad deberás resolverla de manera individual.**

**Nombre y matrícula:**

#### Hiram Garcia Austria
#### A00378771


In [79]:
# Incluye aquí todas las librerías y paquetes que requieras.
import numpy as np
import pandas as pd

from sklearn.decomposition import TruncatedSVD

### **Liga de Datos de la UCI "Restaurant & consumer data".**
### **Descarga el archivo "restaurant+consumer+data.zip" y de ahí necesitaremos los archivos "rating_final.csv" y "geoplaces2.csv".**

https://archive.ics.uci.edu/dataset/232/restaurant+consumer+data



In [80]:
# Descarga los archivos de la página de la UCI a partir de los cuales generaremos
# nuestras matrices de utilidad.

data1 = pd.read_csv("./rating_final.csv", header='infer', sep=",")
data2 = pd.read_csv("./geoplaces2.csv", header='infer',  encoding='latin-1')

print(data1.shape, data2.shape)


(1161, 5) (130, 21)


**NOTA: Recuerda que cada vez que se te pida incluir líneas de código, podrás incluir 1 o más líneas, pero siempre en la sección donde se te indica hacerlo.**

## **Ejercicio - 1**

### **Explica cuál es el propósito de usar el argumento "latin-1" al cargar el segundo archivo.**

++++++++ Inicia la sección de agregar texto: +++++++++++

La palabra encoding significa codificación, y se refiere a la forma en que los caracteres son representados en la memoria de la computadora.

El parámetro "latin-1" es un encoding que se usa para caracteres especiales en español, como la ñ o los acentos.

El encoding por defecto de Python es UTF-8, pero debido a que el archivo "geoplaces2.csv" está codificado en "latin-1", debemos especificarlo para que Python mediante la librería pandas pueda leerlo correctamente.

De no hacerlo, pandas puede arrojar un error al leer el archivo.

++++++++ Termina la sección de agregar texto. +++++++++++

In [81]:
# Del primer archivo obtenemos una matriz con 3 evaluaciones de los restaurantes:
# general, comida y servicio.
# Las evaluaciones pueden ser 0, 1 o 2, siendo 0 la menor calificación y 2 la
# mayor calificación:

data1.head()

userID  placeID  rating  food_rating  service_rating
0  U1077   135085       2            2               2
1  U1077   135038       2            2               1
2  U1077   132825       2            2               2
3  U1077   135060       1            2               2
4  U1068   135104       1            1               2

In [82]:
# Del segundo archivo obtenemos información diversa de cada restaurante:

data2.head(2)

placeID   latitude   longitude  \
0   134999  18.915421  -99.184871   
1   132825  22.147392 -100.983092   

                                      the_geom_meter             name  \
0  0101000020957F000088568DE356715AC138C0A525FC46...  Kiku Cuernavaca   
1  0101000020957F00001AD016568C4858C1243261274BA5...  puesto de tacos   

                                  address        city    state country fax  \
0                              Revolucion  Cuernavaca  Morelos  Mexico   ?   
1  esquina santos degollado y leon guzman      s.l.p.   s.l.p.  mexico   ?   

   ...            alcohol smoking_area dress_code     accessibility   price  \
0  ...  No_Alcohol_Served         none   informal  no_accessibility  medium   
1  ...  No_Alcohol_Served         none   informal        completely     low   

                     url Rambience franchise    area other_services  
0  kikucuernavaca.com.mx  familiar         f  closed           none  
1                      ?  familiar         f    open           none  

[2 rows x 21 columns]

In [83]:
# De data1 no requeriremos "rating" y de data2 solo necesitamos "placeID" y "name":
# Definimos la lista y matrices con los factores que necesitamos:

lista_data1 = ['userID','placeID','food_rating','service_rating']
lista_data2 = ['placeID','name']

data1a = data1[lista_data1]
data2a = data2[lista_data2]

## **Ejercicio - 2**

In [84]:
# Define el DataFrame que conjunte la información de las dos DataFrame data1a
# y data2a en uno solo, a través de la columna común "placeID" y que
# llamaremos "df_combinado":



# ************* Incluye aquí tu código:*****************************


df_combinado = pd.merge(data1a, data2a, on='placeID', how='inner')


# *********** Aquí termina la sección de agregar código *************


# Despleguemos la dimensión y los primeros renglones de este DataFrame:

print(df_combinado.shape)
df_combinado.head()

(1161, 5)


userID  placeID  food_rating  service_rating                     name
0  U1077   135085            2               2   Tortas Locas Hipocampo
1  U1077   135038            2               1    Restaurant la Chalita
2  U1077   132825            2               2          puesto de tacos
3  U1077   135060            2               2  Restaurante Marisco Sam
4  U1068   135104            1               2                     vips

## **Ejercicio - 3**

In [85]:
#    Define la matriz de utilidad cuyos renglones sean los nombres de los
#    restaurantes, las columnas los IDs de los usuarios y las entradas la
#    evaluación de la comida (food_rating). La llamaremos "UtMx_food".


# ************* Incluye aquí tu código:*****************************


UtMx_food = df_combinado.pivot_table(values='food_rating', index='name', columns='userID', fill_value=0)


# *********** Aquí termina la sección de agregar código *************


print('Dimensión de la matriz de Utilidad sobre la evaluación de la comida:')
print('(restaurantes, usuarios) =', (UtMx_food.shape))
UtMx_food.head()

Dimensión de la matriz de Utilidad sobre la evaluación de la comida:
(restaurantes, usuarios) = (129, 138)


userID                     U1001  U1002  U1003  U1004  U1005  U1006  U1007  \
name                                                                         
Abondance Restaurante Bar    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
Arrachela Grill              0.0    0.0    0.0    0.0    0.0    0.0    0.0   
Cabana Huasteca              0.0    0.0    2.0    0.0    0.0    0.0    0.0   
Cafe Chaires                 0.0    0.0    0.0    0.0    0.0    0.0    1.0   
Cafeteria cenidet            0.0    0.0    0.0    0.0    0.0    0.0    0.0   

userID                     U1008  U1009  U1010  ...  U1129  U1130  U1131  \
name                                            ...                        
Abondance Restaurante Bar    1.0    0.0    0.0  ...    0.0    0.0    0.0   
Arrachela Grill              0.0    0.0    0.0  ...    0.0    0.0    0.0   
Cabana Huasteca              0.0    2.0    0.0  ...    0.0    0.0    0.0   
Cafe Chaires                 0.0    0.0    0.0  ...    0.0    0.0    0.0   
Cafeteria cenidet            0.0    0.0    0.0  ...    0.0    0.0    0.0   

userID                     U1132  U1133  U1134  U1135  U1136  U1137  U1138  
name                                                                        
Abondance Restaurante Bar    0.0    0.0    0.0    0.0    0.0    0.0    0.0  
Arrachela Grill              0.0    0.0    0.0    0.0    0.0    0.0    0.0  
Cabana Huasteca              0.0    0.0    1.0    0.0    0.0    0.0    0.0  
Cafe Chaires                 0.0    0.0    0.0    0.0    0.0    0.0    0.0  
Cafeteria cenidet            0.0    0.0    0.0    0.0    0.0    0.0    0.0  

[5 rows x 138 columns]

## **Ejercicio - 4**

### **En la Factorización SVD la cantidad de valores singulares será menor o igual al menor valor de los renglones o columnas de la matriz de utilidad UtMx.**

### **La factorización SVD de una matriz $A$ tiene la forma:**

$A_{m\times n} = U_{m\times m}\Sigma_{m\times n}V_{n\times n}^T$

In [86]:
# a. Aplica la transformación SVD para obtener la matriz de las variables
#    latentes de los restaurantes, en relación a la manera en que los usuarios 
#    evaluaron la comida. Deberás obtener la factorización SVD con el máximo número
#    de componentes y por lo tanto el máximo de valores singulares posibles de acuerdo
#    a este criterio. Así, de acuerdo al resultado obtenido en el Ejercicio-3, asigna
#    a la variable "nc_food" el número entero adecuado para llegar a obtener el 
#    resultado pedido.


# ************* Incluye aquí tu código:**********************************************


nc_food = 128     # número de componentes


# *********** Aquí termina la sección de agregar código *****************************

print('Total de valores singulares basados en evaluación de la comida:', nc_food)

Total de valores singulares basados en evaluación de la comida: 128


In [87]:
# b. Lleva a cabo la factorización SVD truncada en la matriz de utilidad
#    obtenida previamente y con respecto al número de componentes dada 
#    por la variable nc_food.


# ************* Incluye aquí tu código:********************************


SVD_food = TruncatedSVD(n_components=nc_food, random_state=1)
SVD_food.fit(UtMx_food)


# *********** Aquí termina la sección de agregar código ***************


# Con la factorización SVD truncada obtenida, determinemos la cantidad 
# de componentes que explican un 90% de la variabilidad acumulada de 
# cada componente:

for j in range(nc_food):
  if SVD_food.explained_variance_ratio_[0:j].sum() > 0.90:
    break

# Usaremos esta cantidad de componentes para las recomendaciones basadas
# en la calificación de la comida:
N_food = j-1


print('Valor de truncamiento al 90% de dicha variabilidad:', N_food)

Valor de truncamiento al 90% de dicha variabilidad: 51


In [88]:
# Usando la métrica de correlación de Pearson, obtenemos ahora las 10 mejores
# recomendaciones de este modelo no supervisado con base a la información de
# alguien que desea obtener las similitudes con el "Restaurante Pueblo Bonito" 
# y con la cota de truncamiento obtenido en el ejercicio anterior.
# Para ello obtenemos las 10 mejores correlaciones positivas a continuación.

# Factorización SVD:
SVD_food = TruncatedSVD(n_components = N_food)
resultant_matrix_food = SVD_food.fit_transform(UtMx_food)

# Matriz de correlación de Pearson:
corr_mat_food = np.corrcoef(resultant_matrix_food)

# Restaurante de referencia:
restaurante_de_referencia = "Restaurante Pueblo Bonito"
nombres_rest = UtMx_food.T.columns  # nombres de restaurantes
idx_rest = list(nombres_rest).index(restaurante_de_referencia)
corr_rest = corr_mat_food[idx_rest] # Vector de Correlación del RinconHuasteco

# Buscando las correlaciones positivas:
idx = (corr_rest>0)
mejores_sim_food = list()
for i in range(len(nombres_rest[idx])):
  mejores_sim_food.append((corr_rest[idx][i], nombres_rest[idx][i]))

print('Total de similaridades positivas encontradas:', len(mejores_sim_food))

print('Algunos de los resultados encontrados:')
mejores_sim_food[0:5]

Total de similaridades positivas encontradas: 66
Algunos de los resultados encontrados:


[(0.049834158449219415, 'Cabana Huasteca'),
 (0.019160244031031604, 'Cafe Chaires'),
 (0.007549663439215647, 'Cafeteria cenidet'),
 (0.239817285236685, 'Cafeteria y Restaurant El Pacifico'),
 (0.027824224985790708, 'Cenaduria El Rincón de Tlaquepaque')]

## **Ejercicio - 5**

In [89]:
# Ordenar la lista de recomendaciones "mejores_sim_food" encontrada en el paso
# anterior de manera descendente. Llamaremos "mejores_sim_food_ordenadas" a 
# dicha lista.

# ************* Incluye aquí tu código:*****************************



mejores_sim_food_ordenadas = sorted(mejores_sim_food, key=lambda x: x[0], reverse=True)



# *********** Aquí termina la sección de agregar código *************


# Desplegamos las 10 mejores similitudes encontradas de manera descendente:
print('Similitudes con base a la evaluación de la comida con mayores valores de correlación:')
for k in range(1,11):
  print('%d> %s' % (k, mejores_sim_food_ordenadas[k]))

Similitudes con base a la evaluación de la comida con mayores valores de correlación:
1> (0.7922447104880265, 'Restaurante la Estrella de Dima')
2> (0.7090601117958645, 'pizza clasica')
3> (0.6556046736281306, 'Restaurante Guerra')
4> (0.556383654502525, 'El Club')
5> (0.3821658858106644, 'Pizzeria Julios')
6> (0.3140987725684201, 'Hamburguesas saul')
7> (0.304151138469473, 'Restaurant Oriental Express')
8> (0.2930858528705122, 'La Fontana Pizza Restaurante and Cafe')
9> (0.29303747886798687, 'la Cochinita Pibil Restaurante Yucateco')
10> (0.2912676646216186, 'Restaurante El Cielo Potosino')


## **Ejercicio - 6**

### **De manera análoga y usando ahora la evaluación de un restaurante con respecto a su servicio (service_rating), encuentra ahora los diez restaurantes con mayor similaridad al mismo restaurante anterior llamado "Restaurante Pueblo Bonito" y con respecto también a la misma métrica de similaridad de Pearson.**

### **Deberás usar la cantidad de componentes necesarias para que la varabilidad de la factorización SVD quede explicada en un 90%.**

### **a.  Para ello define primero la matriz de utilidad, UtMx_service, donde los renglones son los nombres de los restaurantes, las columnas los usuarios y las entradas las evaluaciones con respecto al servicio recibido en el restaurante.**





In [90]:
# a. Define la matriz de utilidad cuyos renglones sean los nombres de los
#    restaurantes, las columnas los IDs de los usuarios y las entradas la
#    evaluación del servicio (service_rating). La llamaremos "UtMx_service".


# ************* Incluye aquí tu código:*****************************


UtMx_service = df_combinado.pivot_table(values='service_rating', index='name', columns='userID', fill_value=0)


# *********** Aquí termina la sección de agregar código *************




print('Dimensión de la matriz de Utilidad sobre la evaluación del servicio:')
print('(restaurantes, usuarios) =', (UtMx_service.shape))
UtMx_service.head()

Dimensión de la matriz de Utilidad sobre la evaluación del servicio:
(restaurantes, usuarios) = (129, 138)


userID                     U1001  U1002  U1003  U1004  U1005  U1006  U1007  \
name                                                                         
Abondance Restaurante Bar    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
Arrachela Grill              0.0    0.0    0.0    0.0    0.0    0.0    0.0   
Cabana Huasteca              0.0    0.0    2.0    0.0    0.0    0.0    0.0   
Cafe Chaires                 0.0    0.0    0.0    0.0    0.0    1.0    1.0   
Cafeteria cenidet            0.0    0.0    0.0    0.0    0.0    0.0    0.0   

userID                     U1008  U1009  U1010  ...  U1129  U1130  U1131  \
name                                            ...                        
Abondance Restaurante Bar    1.0    0.0    0.0  ...    0.0    0.0    0.0   
Arrachela Grill              0.0    0.0    0.0  ...    0.0    0.0    0.0   
Cabana Huasteca              0.0    1.0    0.0  ...    0.0    0.0    0.0   
Cafe Chaires                 0.0    0.0    0.0  ...    0.0    0.0    0.0   
Cafeteria cenidet            0.0    0.0    0.0  ...    0.0    0.0    0.0   

userID                     U1132  U1133  U1134  U1135  U1136  U1137  U1138  
name                                                                        
Abondance Restaurante Bar    0.0    0.0    0.0    0.0    0.0    0.0    0.0  
Arrachela Grill              0.0    0.0    0.0    0.0    0.0    0.0    0.0  
Cabana Huasteca              0.0    0.0    1.0    0.0    0.0    0.0    0.0  
Cafe Chaires                 0.0    0.0    0.0    0.0    0.0    0.0    0.0  
Cafeteria cenidet            0.0    0.0    0.0    0.0    0.0    0.0    0.0  

[5 rows x 138 columns]

In [91]:
# b. Incluye el código necesario para obtener y desplegar los 10 restaurantes con
#    mayor similaridad de Pearson al "Restaurante Pueblo Bonito", y usando solo
#    el 90% de las componentes que explican la varianza de la matriz UtMx_service.
#    Desplegarlos en orden descendente, sin incluir el restaurante de referencia.

# ************* Incluye aquí tu código:*****************************

# número de componentes
nc_service = 128
SVD_service = TruncatedSVD(n_components=nc_service, random_state=1)
SVD_service.fit(UtMx_service)

for j in range(nc_service):
  if SVD_service.explained_variance_ratio_[0:j].sum() > 0.90:
    break

# Usaremos esta cantidad de componentes para las recomendaciones basadas
# en la calificación de la comida:
N_service = j-1

print('Valor de truncamiento al 90% de dicha variabilidad:', N_service)



# *********** Aquí termina la sección de agregar código *************

Valor de truncamiento al 90% de dicha variabilidad: 51


In [92]:
# Factorización SVD:
SVD_service = TruncatedSVD(n_components = N_service)
resultant_matrix_service = SVD_service.fit_transform(UtMx_service)

# Matriz de correlación de Pearson:
corr_mat_service = np.corrcoef(resultant_matrix_service)

# Restaurante de referencia:
restaurante_de_referencia = "Restaurante Pueblo Bonito"
nombres_rest = UtMx_service.T.columns  # nombres de restaurantes
idx_rest = list(nombres_rest).index(restaurante_de_referencia)
corr_rest = corr_mat_service[idx_rest] # Vector de Correlación

# Buscando las correlaciones positivas:
idx = (corr_rest>0)
mejores_sim_service = list()
for i in range(len(nombres_rest[idx])):
  mejores_sim_service.append((corr_rest[idx][i], nombres_rest[idx][i]))

print('Total de similaridades positivas encontradas:', len(mejores_sim_service))

print('Algunos de los resultados encontrados:')
mejores_sim_service[0:5]

Total de similaridades positivas encontradas: 57
Algunos de los resultados encontrados:


[(0.23345834199723683, 'Abondance Restaurante Bar'),
 (0.02302107593586922, 'Cabana Huasteca'),
 (0.17830286935304898, 'Cafeteria y Restaurant El Pacifico'),
 (0.04065673449687551, 'Carls Jr'),
 (0.03251978081483441, 'Cenaduria El Rincón de Tlaquepaque')]

In [93]:
# Ordenar la lista de recomendaciones 
mejores_sim_service_ordenadas = sorted(mejores_sim_service, key=lambda x: x[0], reverse=True)

# Desplegamos las 10 mejores similitudes encontradas de manera descendente:
print('Similitudes con base a la evaluación del servicio con mayores valores de correlación:')

for k in range(1,11):
    print('%d> %s' % (k, mejores_sim_service_ordenadas[k]))


Similitudes con base a la evaluación del servicio con mayores valores de correlación:
1> (0.6636178140925859, 'pizza clasica')
2> (0.5815825672397777, 'El Club')
3> (0.4781169028890966, 'Restaurante la Estrella de Dima')
4> (0.4744595254922273, 'Hamburguesas saul')
5> (0.4302479779343318, 'Restaurante Guerra')
6> (0.41637112615747657, 'Restaurant Orizatlan')
7> (0.36346937154426595, 'la Cochinita Pibil Restaurante Yucateco')
8> (0.3515021577782078, 'Restaurante El Cielo Potosino')
9> (0.3116277370131301, 'La Fontana Pizza Restaurante and Cafe')
10> (0.30177860534939516, 'puesto de tacos')


## **Ejercicio - 7**

### **Incluye tus comentarios y conclusiones de la actividad. En particular indica cuáles restaurantes encontraste con correlaciones altas tanto para la calificación de la comida, como del servicio.**

++++++++ Inicia la sección de agregar texto: +++++++++++

### **Comentarios:**
Esta ha sido una actividad muy interesante en la que hemos aplicado la factorización de matrices a un problema de recomendación de restaurantes. Hemos obtenido las 10 mejores recomendaciones basadas en la evaluación de la comida y del servicio, con respecto a un restaurante de referencia.

Me ha sorprendido la facilidad con la que se puede llevar a cabo la factorización SVD truncada y obtener las recomendaciones de manera muy rápida. Además, la posibilidad de obtener las recomendaciones basadas en la evaluación de la comida y del servicio, me parece una excelente manera de ofrecer un servicio personalizado a los usuarios.

Agradezco por la oportunidad de aprender y aplicar estos conceptos en un problema real de recomendación de restaurantes. Y aun más que yo siendo de Cuernavaca, me ha sido muy grato ver lugares conocidos estando yo viviendo en Israel.

### **Resultados:**
Las recomendaciones basadas en la evaluación de la comida y del servicio, con respecto al "Restaurante Pueblo Bonito", han sido muy interesantes. En ambos casos, se han encontrado restaurantes con evaluaciones muy similares a las del restaurante de referencia. Esto me hace pensar que las recomendaciones basadas en la evaluación de la comida y del servicio son muy acertadas.

En el primer caso, las dos recomendaciones más altas basadas en la **evaluación de la comida** para alguien que le guste el "Restaurante Pueblo Bonito", son: "Restaurante la Estrella de Dima" (con 79.22%) y "pizza clásica" (con 70.90%)

En el segundo caso, las dos recomendaciones más altas basadas en la **evaluación del servicio** para alguien que le guste el "Restaurante Pueblo Bonito", son: "pizza clásica" (con 66.36%) y "El Club" (con 58.15%)

El porcentaje de correlación es un valor que va de 0 a 100, donde 100 es una correlación perfecta. Por lo que podemos decir que las recomendaciones respecto a la comida son bastante acertadas, mientras que las recomendaciones respecto al servicio son un poco menos acertadas.


++++++++ Termina la sección de agregar texto. +++++++++++

>> **Fin de la Actividad de Sistemas de Recomendación**